src/kernel/page.c

```diff
@@ -194,6 +194,19 @@ boolean validate_virtual(void * base, u64 length)
     return traverse_ptes(u64_from_pointer(base), length, stack_closure(validate_entry));
 }
 
+closure_function(0, 3, boolean, validate_entry_writable,
+                 int, level, u64, vaddr, pteptr, entry)
+{
+    pageflags pf = pageflags_from_pteptr(entry);
+    return pageflags_is_present(pf) && pageflags_is_writable(pf);
```

Comment:
Why not just use pte_is_present here? Also, is it true on every arch that every level must be writable, or just the leaf? I think that's the case, but not totally sure.

Reply:
You're right, this should only be testing the leaves for writability. I rewrote the expression accordingly.